Make `aiod_entry.status` a true ENUM

PGijsbers opened the pull request:
Please have a look at this proposal and see if it seems reasonable. I will ping for a proper review later after collecting feedback. It may be better to look at the two commits separately.

## Change
<!-- Briefly describe the change of this PR -->
Previously, `aiod_entry.status` was a [NamedRelation](https://github.com/aiondemand/AIOD-rest-api/blob/develop/src/database/model/named_relation.py). In this PR, we update it to be an [enum](https://github.com/aiondemand/AIOD-rest-api/blob/6038b0c0e8c359a1a5c8275e86fa60dbd958e4b4/src/database/model/concept/aiod_entry.py#L25) instead.

The main benefit is that we can now from the source code determine the possible values, which means that we can refer to it with the Enum. This makes errors much less error prone, and enables selection of possible values without knowing what is stored in the database. For a type which is rarely (if at all) subject to change, this seems appropriate.
 
It also enforces the constraints on the values the `status` can have, which is something we need and NamedRelation currently does not. Though it should be noted that in the future we probably want to support enforcing constraints on (some/most?) NamedRelations anyway. 


If the general direction is accepted, we still need to:
 - [x] Add a migration script to remove the old tables from the database, and replace the `status` column type and its foreign key with the direct enum values.
 - [x] Figure out if/how to update Elastic Search indices (@AdrianAlcolea, do you know? Or should we simply delete the indices and recreate?)

## How to Test
<!-- Describe a way to test the change of this PR -->

## Checklist
- [x] Tests have been added or updated to reflect the changes, or their absence is explicitly explained.
- [x] Documentation has been added or updated to reflect the changes, or their absence is explicitly explained.
- [x] A self-review has been conducted checking:
  - No unintended changes have been committed.
  - The changes in isolation seem reasonable.
  - Anything that may be odd or unintuitive is provided with a GitHub comment explaining it (but consider if this should not be a code comment or in the documentation instead).
- [x] All CI checks pass before pinging a reviewer, or provide an explanation if they do not.

## Related Issues
<!-- Provide a list of relevant issues and/or pull requests, if any. -->
<!-- If the pull request closes and issue, specify "Closes #X" (where X is the issue number). -->

PGijsbers committed:
Force draft status to be a regular enum instead of namedrelation

PGijsbers committed:
Add entry status to the elastic search indices

Taniya-Das reviewed (approved):
The proposed changes look good to me. The constraints on the status field make more sense than how it was previously defined.

AdrianAlcolea commented:
I think is a good idea (for the moment I didn't review al the modification in the first commit, but i like the concept itself).
Regarding the elasticsearch service, I need to check a bit more. I think that adding new field in the mappings if feasible without recreating, but I don't know how to actualise the content itself. (Or we can just recreate).
In any case, the thing is that I'm not sure if it's needed to add the status in elasticsearch indices (we didn't have it before). The approach was to only index the text fields that are susceptible of being used for the search (name, description, and some specific fields like ISBN etc...) and just add that to elasticsearch to keep the search light. Then, there is the "get_all" field to retrieve all the content, in which case we made one access to elastisearch and then use the indices to access the database and retrieve everything. We should only index it if we want to add a field in the search endpoints to select the status and retrieve only documents with a particular status.
If this is the case then the second commit is ok too, but we also need to add it in setup/es_setup/definitions.py and in the global_indexed_fields of the search_router.

PGijsbers commented:
@AdrianAlcolea Thanks. I think this largely depends on whether or not we want to support asset visibility in Elastic Search. In the future, people will be able to have 1) published assets that are private and 2) assets in draft status or under review, both of these should only be visible to users that have read access to the asset. Without including the visibility and asset status in the search index, I do not think we can support searching through _all_ assets (including private ones) and returning only those to which the user has access in a performant manner. I feel that we have two choices here:

1. Do not index anything "private" at all. Only "published" assets which are not marked as private. This means that we do not have to make any further modifications to the search API (though we do need to update the WHERE clause of ES updates), but means that a user querying the API will never see its own private assets either in the ES endpoints. This might not be a big deal (until it is, at which point adopting the alternative would probably also not be that costly).
1. Index information about the asset status and visibility settings, but index all data (including private data). In this case, we will have to extend the search endpoint to filter out assets from the results to which the requesting user does not have access. This adds some complexity, especially when we have more flexible permissions (e.g., groups that can see a private asset, instead of just the uploader).


So while adding the entry status was deliberate, and needed for no2, I think going with option no1 might be better at this point. We would update the WHERE clauses of the queries to take into asset status (and later visibility), thus not index private data at all. This significantly reduces the complexity, without really hindering us should we later want to support scenario no1 after all.

PGijsbers committed:
Revert "Add entry status to the elastic search indices"

This reverts commit 6038b0c0e8c359a1a5c8275e86fa60dbd958e4b4.

AdrianAlcolea commented:
Great, I also think that first approach is the best one for the moment. If I understood we will add a WHERE clause to only select assets with 'published' status right? And once the 'visibility' is implemented also take only the 'public' ones.
In the next release or update of production, with this modification already implemented, we should also manually remove the assets already indexed as 'draft'.
This is ok for me =)

PGijsbers commented:
I think all assets currently listed as draft should just be marked "published" since there is currently no review process//real concept of draft vs published. everything is visible immediately. other than that, yes pretty much as you say 👍

PGijsbers committed:
Add convenience script to allow setting alembic revision

PGijsbers committed:
Add migration script to make Entry Status an Enum

PGijsbers commented on alembic/alembic/versions/1662d64ebe23_make_draft_status_enum.py:
```diff
@@ -0,0 +1,90 @@
+
+
+def upgrade() -> None:
+    op.drop_table("aiod_entry_status_link")
```
The `status` table had a key constraint to this one, so we cannot drop it without at least removing the key constraint. This table was seemingly introduced by error (I checked with Jos), it would support a many-to-many relation between status and entry, which is not something that was ever supported in the API. So we simply drop this table.

PGijsbers commented:
I added the migration script. You should now be able to create a database on the `develop` branch, populate it, move to `enh/entry_status` branch, invoke the migration script (if it gives an error about news source, use `./scripts/set_alembic.sh d09ed8ad4533` to make Alembic aware which revision the database is on, and then start migration again), and both the data and schema should be transferred over fully. 
Note that the identifiers for different statuses depend on the order of insertion in the database. E.g., if the first asset you create on your fresh database is a "published" asset, then the "published" status will correspond to "status_identifier=1". In the enum mapping, I used the mapping as it is on the production and test databases. 
It is also possible to undo this migration and re-introduce the status table by doing an alembic downgrade to d09ed8ad4533. @Taniya-Das Since you had a closer look at the schema changes before already, would you mind having a look/test at the migration I added? Thanks a bundle!

@AdrianAlcolea I reverted the ES changes, and will instead add them in another PR. This PR now only takes care of the schema migration without affecting the current behavior around visibility/searchability of assets depending on entry status.

Taniya-Das commented:
I have tested it and the changes look good to me.
